staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go

```diff
@@ -745,3 +746,121 @@ func determineRequestedIP(svc *v1.Service, fwdRule *compute.ForwardingRule) stri
 
 	return ""
 }
+
+func getILBOptions(svc *v1.Service) *ILBOptions {
+	globalAccess := GetLoadBalancerAnnotationGlobalAccess(svc)
+	if !globalAccess {
+		return nil
+	}
+	return &ILBOptions{EnableGlobalAccess: globalAccess}
+}
+
+type fwdRuleIntf struct {
+	name            string
+	betaDescription string
+	gaDescription   string
+	ipAddress       string
+	backendService  string
+	ports           []string
+	ipProtocol      string
+	lbScheme        string
+	subnetwork      string
+	network         string
+	gaRule          *compute.ForwardingRule
+	betaRule        *computebeta.ForwardingRule
+}
+
+func (f *fwdRuleIntf) constructRules(options *ILBOptions) {
+	f.gaRule = &compute.ForwardingRule{
+		Name:                f.name,
+		Description:         f.gaDescription,
+		IPAddress:           f.ipAddress,
+		BackendService:      f.backendService,
+		Ports:               f.ports,
+		IPProtocol:          f.ipProtocol,
+		LoadBalancingScheme: f.lbScheme,
+	}
+	f.betaRule = &computebeta.ForwardingRule{
+		Name:                f.name,
+		Description:         f.betaDescription,
+		IPAddress:           f.ipAddress,
+		BackendService:      f.backendService,
+		Ports:               f.ports,
+		IPProtocol:          f.ipProtocol,
+		LoadBalancingScheme: f.lbScheme,
+	}
+	if f.subnetwork != "" {
+		f.gaRule.Subnetwork = f.subnetwork
+		f.betaRule.Subnetwork = f.subnetwork
+	} else {
+		f.gaRule.Network = f.network
+		f.betaRule.Network = f.network
+	}
+	if options != nil {
+		f.betaRule.AllowGlobalAccess = options.EnableGlobalAccess
+	}
+}
+
+func getFwdRuleAPIVersions(rule *compute.ForwardingRule, options *ILBOptions) (oldVersion, newVersion string) {
+	if options != nil && options.EnableGlobalAccess {
+		newVersion = computeBetaVersion
+	}
+	values := make(map[string]string)
+	if rule != nil {
+		json.Unmarshal([]byte(rule.Description), &values)
+		oldVersion = values[apiVersionString]
+	}
+	return oldVersion, newVersion
+}
+
+// fwdRuleCompare returns true if the rules are different, returns false if the rules are identical.
+func (g *Cloud) fwdRuleCompare(existingRule *compute.ForwardingRule, expectedRule *fwdRuleIntf, options *ILBOptions, version string) (bool, error) {
+	if version == "" {
+		return !fwdRuleGAEqual(existingRule, expectedRule.gaRule), nil
+	}
+	existingBetaRule, err := g.GetBetaRegionForwardingRule(existingRule.Name, g.region)
+	if err != nil {
+		// cannot fetch rule, assume the rules are different
+		klog.Errorf("ensureInternalLoadBalancer(%s): failed to retrieve existing forwarding rule - %s", existingRule.Name, err)
+		return true, err
+	}
+	return !fwdRuleBetaEqual(existingBetaRule, expectedRule.betaRule), nil
+}
+
+func (g *Cloud) ensureInternalForwardingRule(existingFwdRule *compute.ForwardingRule, expectedRuleIntf *fwdRuleIntf, options *ILBOptions) (err error) {
+	expectedRuleIntf.constructRules(options)
+	lbName := expectedRuleIntf.name
```

Comment:
looks like this is just for logging, I would prefer just using `expectedRuleIntf.name`

Reply:
done